pandas/core/case_when.py

```diff
@@ -0,0 +1,91 @@
+from __future__ import annotations
+
+from typing import (
+    Any,
+    Callable,
+)
+
+import numpy as np
+
+from pandas._libs import lib
+
+import pandas as pd
+import pandas.core.common as com
+
+
+def case_when(*args, default: Any = lib.no_default) -> Callable:
```

Comment:
Users wanting to use this without assign would be calling something like

    pd.case_when(cond1, val1, cond2, val2)(df)

That seems odd, and it's much more natural to have the DataFrame as an argument. On the other hand, that wouldn't allow as much usage with assign via method chaining if the DataFrame is changed as part of the chain.

Reply:
I think it should return a series instead

Reply:
> On the other hand, that wouldn't allow as much usage with assign via method chaining if the DataFrame is changed as part of the chain.

Exactly, what do you suggest in this case?

Reply:
I don't think we should start adding functions to the API just to support method chaining in assign. Users can still use this with a simple wrapper, e.g.

```
def chain_case_when(df, *args, **kwargs):
    return lambda df: pd.case_when(df, *args, **kwargs)

result = (
    ...
    .assign(new_col=chain_case_when(lambda x: x.a == 1, 'first', lambda x: (x.a > 1), 'second', default='default'))
)
```

A few other thoughts on the API. A few of these repeat other comments here, just adding so it's all in one place.

 - I think the `default` argument should have no default (must be user-specified) https://github.com/pandas-dev/pandas/pull/50343#discussion_r1064136418
 - Return value should be pandas object (Series)
 - The top level method is necessary for constructing a Series where the default is a scalar. However it also seems natural to also have `Series.case_when` where `default` is the provided Series (so not an argument in the signature). We could also have `DataFrame.case_when` be an alias for `pd.case_when(df, ...)`.
 - Needs to support EAs (https://github.com/pandas-dev/pandas/pull/50343#discussion_r1064051279)
 - Does the current implementation support replacements (values) being Series? I think the answer is yes, but wanted to be sure. This should be added to docs and tests.

The tests should also be expanded:
 - condition being the Boolean EA
 - dtypes for values including bool, string, and EAs (numeric, categorical, datetimelike, period)
 - Malformed arguments. E.g. what happens when a condition is something unexpected like a scalar or a list / ndarray / Series that is the wrong length or not aligned?

Reply:
I personally feel pd.Series.mask should be used instead of np.select. the implementation of pd.Series.mask takes care of a lot of things, including dtypes and alignment. 

Having case_when as a function makes it versatile, one can use it within assign or on its own. Agreed that it should return a pandas object, preferably a series.

Reply:
Agreed on using mask (or mask-like internals if there is unnecessary logic that is wasting cycles)

> Having case_when as a function makes it versatile

Just for clarification, is this opposition to also having `Series.case_when`?

Reply:
Since pd.case_when returns a pandas object I don't see a need for series.case_when( although it is convenient).

Reply:
Agreed it's not logically necessary, but it seems to be a common use case and using the a top level function is unnecessarily redundant:

    # If case_when supports Series, then can remove the `.to_frame()`
    pd.case_when(ser.to_frame(), cond1, val1, cond2, val2, default=ser)

as opposed to

    ser.case_when(cond1, val1, cond2, val2)

Reply:
Kindly see https://github.com/pandas-dev/pandas/pull/50343#issuecomment-1460350497